### PR TITLE
add 3 commands to motoros and increase pulse thresh

### PR DESCRIPTION
**Pull request by rdiankov**
add following 3 commands:
-    ROS_CMD_START_SERVOS = 200112, // starts the servo motors
-    ROS_CMD_STOP_SERVOS = 200113, // stops the servo motors and motion
-    ROS_CMD_RESET_ALARM = 200114, // clears the error in the current controller

also increase pulse threshold since on mh24 i've seen the 4th axis deviate by ~55 pulses

and add dnProj file for DX200 controller

### CODE REVIEW

File: motoman_driver/MotoPlus/MotionServer.c [hunk 1]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+//-----------------------------------------------------------------------
+BOOL Ros_MotionServer_StartServos(Controller* controller)
+{
+    int ret;
+	// Servo On
+	if(Ros_Controller_IsServoOn(controller) == FALSE)
+	{
+        printf("starting servos\n");
+		MP_SERVO_POWER_SEND_DATA sServoData;
+     	MP_STD_RSP_DATA rData;
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:708]: missing indent?

Or seems spaces vs tabs.

File: motoman_driver/MotoPlus/MotionServer.c [hunk 2]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+			Ros_Controller_ErrNo_ToString(rData.err_no, errMsg, ERROR_MSG_MAX_SIZE);
+			printf("Can't turn on servo because: %s\r\n", errMsg);
+			goto updateStatus;			
+		}
+	}
+	
+updateStatus:	
+	// Update status
+	Ros_Controller_StatusUpdate(controller);
+    return Ros_Controller_IsServoOn(controller) == TRUE;
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:743]: spaces vs tabs

File: motoman_driver/MotoPlus/MotionServer.c [hunk 3]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+}
 
+//-----------------------------------------------------------------------
+// Stops the servos and makes sure that the motion is also stopped
+//-----------------------------------------------------------------------
+BOOL Ros_MotionServer_StopServos(Controller* controller)
+{
+    int ret;
+    Ros_MotionServer_StopMotion(controller);
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:752]: tabs / spaces.

File: motoman_driver/MotoPlus/MotionServer.c [hunk 4]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+
+ 	// Servo Off
+	if(Ros_Controller_IsServoOn(controller) == TRUE)
+	{
+        printf("stopping servos\n");
+		MP_SERVO_POWER_SEND_DATA sServoData;
+        MP_STD_RSP_DATA rData;
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:759]: and here (and second `memset(..)` line)

File: motoman_driver/MotoPlus/MotionServer.c [hunk 5]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+			Ros_Controller_ErrNo_ToString(rData.err_no, errMsg, ERROR_MSG_MAX_SIZE);
+			printf("Can't turn off servo because: %s\r\n", errMsg);
+			goto updateStatus;			
+		}
+	}
+	
+updateStatus:	
+	// Update status
+	Ros_Controller_StatusUpdate(controller);
+    return Ros_Controller_IsServoOn(controller) == FALSE;
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:794]: same

File: motoman_driver/MotoPlus/MotionServer.c [hunk 6]
```diff
@@ -655,7 +694,138 @@ BOOL Ros_MotionServer_StopMotion(Controller* controller)
...
+}
+
+BOOL Ros_MotionServer_ResetAlarm(Controller* controller)
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:797]: entire function is space indented.

File: motoman_driver/MotoPlus/MotionServer.c (hunk 3 above)
Comment (ted-miller) [inline motoman_driver/MotoPlus/MotionServer.c:752]: Is the Ros_MotionServer_StopServos function really necessary?  There is a lot of repeat code.  Could you make one function that takes a boolean?  
Ros_MotionServer_ServoControl(Controller\* controller, BOOL bEnable)

File: motoman_driver/MotoPlus/SimpleMessage.h [hunk 7]
```diff
@@ -133,6 +134,9 @@ typedef enum
 	ROS_CMD_CHECK_MOTION_READY = 200101,
 	ROS_CMD_CHECK_QUEUE_CNT = 200102,
 	ROS_CMD_STOP_MOTION = 200111,
+	ROS_CMD_START_SERVOS = 200112, // starts the servo motors
```
Comment (ted-miller) [inline motoman_driver/MotoPlus/SimpleMessage.h:137]: @shaun-edwards 
I know these command id's are motoman specific... but I wonder how the original number scheme was chosen.  Do these correspond to some similar numbering scheme in ROS-I SimpleMessage?
Are these chosen ID's ok?

**Comment by rdiankov** [comment]
that's what happens when i use emacs to edit ;0)
is there some auto white space fixer for the style you guys are using?

**Comment by gavanderhoorn** [comment]
@rdiankov wrote:

> is there some auto white space fixer for the style you guys are using?

don't think so. Major thing is tabs-for-indents, but @ted-miller wrote this.

**Comment by ted-miller** [comment]
Both MotoPlusIDE and Visual Studio used tabs by default.  So, this is the style I use for MotoPlus applications.

I'm not very proficient in emacs, but I found this on google: http://www.emacswiki.org/emacs/IndentationBasics

File: motoman_driver/MotoPlus/SimpleMessage.h (hunk 7 above)
Comment (shaun-edwards) [inline motoman_driver/MotoPlus/SimpleMessage.h:137]: @ted-miller, correct me if I am wrong, but I believe these are "sub-command" types that are under a general Motoman SimpleMessage command ID.  Simple message IDs and ranges are specified [here](https://github.com/ros-industrial/rep/blob/master/rep-I0004.rst).  There isn't any overlap between these sub command types and anything envisioned in the simple message spec.  I believe the number decision was Motoman's.

Comment (ted-miller) [inline motoman_driver/MotoPlus/SimpleMessage.h:137]: Ok.  Sounds good to me.

**Comment by gavanderhoorn** [comment]
Should the design document (M2092-EDS ver1.1.4.pdf, from [here](http://wiki.ros.org/motoman_driver#Design)) be updated after merging this PR?

**Comment by ted-miller** [comment]
Yep.  It’s actually out of date already (before this merge).  I’ll put it on the todo list.

-Ted

**Comment by shaun-edwards** [comment]
So if I follow everything correctly, there are still two open issues with this PR:
- [ ] spacing
- [ ] some repeated code in `Ros_MotionServer_StopServos`

@rdiankov, will you be able to address these?

I'll add an issue for updating the Motoman spec (and assign it to @ted-miller ;), since the spec is already out of date and independent of this PR.

**Comment by ted-miller** [comment]
@shaun-edwards  @gavanderhoorn  
@rdiankov has contacted me directly in regard to this pending PR that is a year old.  His latest updates appear useful.  But, they are out of date by now.
I recommend that you close this PR, without merging the changes.  This week I will merge his code into my local copy and test that everything works in the general case.  I will then open a new PR for the MotoROS driver.

**Comment by gavanderhoorn** [comment]
A new PR is ok with me, but do those changes also include:

> also increase pulse threshold since on mh24 i've seen the 4th axis deviate by ~55 pulses

we've run into that in the APC setup as well, but I'm not entirely convinced that just upping the threshold is the best way to work around / fix this.

And it would be really nice if the changes proposed by @rdiankov could be separate from others in the same PR. That way we can review them separately as well.

Alternatively we could just rebase this PR. That should not be too difficult.

**Comment by ted-miller** [comment]
He implemented a different fix for this issue.  I have only briefly glanced at the updates, but it doesn't appear to be a fixed threshold value.

I want to test his changes locally in my lab to ensure they work for the generalized case.  I see one or two items that might need to be slightly adjusted.  That's why I'm not suggesting rebasing.

**Comment by gavanderhoorn** [comment]
Ok. Sounds reasonable.

If @rdiankov is ok with the loss of attribution, I'm ok with closing this one and waiting for @ted-miller's replacement PR.

**Comment by gavanderhoorn** [comment]
#132 is a replacement for this.

**Pull request closed without merging**